napari/utils/_injection.py

```diff
@@ -35,31 +40,74 @@ def get_accessor(type_: Type[T]) -> Optional[Callable[[], Optional[T]]]:
     `inject_napari_dependencies`, allows us to inject current napari objects
     into functions based on type hints.
     """
-    from ..components import LayerList
-    from ..layers import Layer
-    from ..viewer import Viewer, current_viewer
+    if type_ in _ACCESSORS:
+        return _ACCESSORS[type_]
 
     if isinstance(type_, type):
-        if issubclass(type_, Layer):
-            return _get_active_layer
-        if issubclass(type_, Viewer):
-            return current_viewer
-        if issubclass(type_, LayerList):
-            return _get_active_layer_list
+        for key, val in _ACCESSORS.items():
+            if issubclass(type, key):
+                return val  # type: ignore [return-type]
+    return None
+
+
+class set_accessor:
+    """Set accessor(s) for given type(s).
+
+    "Acessors" are functions that can retrieve an instance of a given type.
+    For instance, `napari.viewer.current_viewer` is a function that can
+    retrieve an instance of `napari.Viewer`.
+
+    This is a class that behaves as a function or a context manager, that
+    allows one to set an accessor function for a given type.
+
+    Parameters
+    ----------
+    mapping : Dict[Type[T], Callable[..., Optional[T]]]
+        a map of type -> accessor function, where each value is a function
+        that is capable of retrieving an instance of the associated key/type.
+    clobber : bool, optional
+        Whether to override any existing accessor function, by default True.
+
+    Raises
+    ------
+    ValueError
+        if clobber is `True` and one of the keys in `mapping` is already
+        registered.
+    """
+
+    def __init__(
+        self, mapping: Dict[Type[T], Callable[..., Optional[T]]], clobber=True
+    ):
+        self._before = {}
+        for k, v in mapping.items():
+            if k in _ACCESSORS and clobber is False:
+                raise ValueError(
+                    f"Class {k} already has an accessor and clobber is False"
+                )
```

Comment:
again just a q for clarity - clobbering will only be used when temporarily overriding?

Reply:
in the PR as it currently stands, the clobber flag doesn't care whether you're using `set_accessor` as a context manager or not.  If you call it not as a context manager, it's permanent